### PR TITLE
Add soft actor critic

### LINKED ISSUE #63: Add new RL algorithm to compare with baseline (SAC Discrete)
Our current baseline RL algorithm is DQN (more accurately it is DDQN). Named algorithm uses epsilon-greedy policies to at least have a chance of fully investigating environment in question. Using epsilon this way also makes us schedule change of this parameter during study and this task, schedule, can be non-trivial. So one can try turn to the soft RL problem allowing to deal with entropy of an output policy. One of the algorithms solving soft RL problem is Soft Actor Critic (SAC). As our environment has discrete action space we are interested in Discrete SAC. The hypothesis is that SAC Discrete can play not worse than the best result of DDQN

### PR DESCRIPTION
**Pull request by Cattharine**
Added Soft Actor Critic Discrete with guides. There are two implementations: with and without learning temperature

### CODE REVIEW

File: algorithms/soft_actor_critic.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+            nn.ReLU(),
+            nn.Linear(self.inner_layer, self.inner_layer),
+            nn.ReLU(),
+            nn.Linear(self.inner_layer, action_n)
+        )
+
+    def forward(self, states):
+        return self.network(states)
+
+    def predict(self, state):
```
Comment (krutovsky-danya) [inline algorithms/soft_actor_critic.py:28]: Тут не нужент torch.no_grad?

File: algorithms/soft_actor_critic.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        # masked_fill is out-of-place version of masked_fill_
+        # (in code it usually looks like this: tensor[mask] = value
+        # where mask is a Tensor of booleans)
+        logits = logits.masked_fill(~guides, -torch.inf)
+        return self.softmax(logits)
+
+    def predict_guided(self, state, guide):
+        logits = super().predict(state)
+        guide = guide.unsqueeze(dim=0)
+        logits[~guide] = -torch.inf
```
Comment (krutovsky-danya) [inline algorithms/soft_actor_critic.py:85]: Почему для однообразия не использовать masked_fill?

File: algorithms/soft_actor_critic.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+    @torch.no_grad()
+    def get_action(self, state, info):
+        mask = info["actions"]
+        mask = self._convert_info(torch.tensor(mask))
+        state = torch.FloatTensor(state).to(device=self.device)
+        masked_probs = self.policy_function.predict_guided(state, mask).squeeze().cpu()
+        # had an error with probs sum not being equal to 1 due to problems with accuracy
+        # and inner implicit casting from float32 to float64
+        np_probs = np.asarray(masked_probs).astype("float64")
```
Comment (krutovsky-danya) [inline algorithms/soft_actor_critic.py]: А ошибка все еще сохранилась?

File: algorithms/soft_actor_critic.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+            torch.stack, list(zip(*batch))
+        )
+
+        states = states.to(self.device)
+        guides = guides.to(self.device)
+        actions = actions.to(self.device)
+        rewards = rewards.to(self.device)
+        dones = dones.to(self.device)
+        next_states = next_states.to(self.device)
+        next_guides = next_guides.to(self.device)
```
Comment (krutovsky-danya) [inline algorithms/soft_actor_critic.py]: Давай вынесем эту логику в отдельный метод? Кажется этот код немного мешает пониманию происходящего в fit

File: algorithms/soft_actor_critic.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        return (loss_policy, *loss_q)
+
+    def _q_gradient_step(self, states, actions, rewards, dones, next_states, next_guides):
+        next_q_target_1 = self.target_q_function_1.forward(next_states)
+        next_q_target_2 = self.target_q_function_2.forward(next_states)
+        next_min_q_target = torch.min(next_q_target_1, next_q_target_2)
+
+        next_action_probs = self.policy_function.forward_guided(next_states, next_guides)
+        next_action_log_probs = get_nan_safe_log(next_action_probs, self.small_const)
+        entropy = self.alpha.detach() * next_action_log_probs
```
Comment (krutovsky-danya) [inline algorithms/soft_actor_critic.py]: Кажется это только часть энтропии, чтобы не путаться я бы предложил назвать переменную по-другому или честно посчитать энтропию

File: algorithms/soft_actor_critic.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+
+        return (loss_policy, *loss_q)
+
+    def _q_gradient_step(self, states, actions, rewards, dones, next_states, next_guides):
+        next_q_target_1 = self.target_q_function_1.forward(next_states)
+        next_q_target_2 = self.target_q_function_2.forward(next_states)
+        next_min_q_target = torch.min(next_q_target_1, next_q_target_2)
+
+        next_action_probs = self.policy_function.forward_guided(next_states, next_guides)
+        next_action_log_probs = get_nan_safe_log(next_action_probs, self.small_const)
```
Comment (krutovsky-danya) [inline algorithms/soft_actor_critic.py]: Можно получить log_probs отнормировав логиты (logits - torch.logsumexp(logits))

File: algorithms/soft_actor_critic.py [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        _take_optimization_step(self.q_optimizer_2, loss_q_2)
+
+        return loss_q_1, loss_q_2
+
+    def _policy_gradient_step(self, states, guides):
+        # учитывая, что loss_policy ─ это сумма с учетом вероятностей действий action_probs
+        # (которые обнулены на недоступных действиях и отнормированы softmax по доступным),
+        # а также то, что градиент в данном случае не будет течь через значения q функции,
+        # не важно, какие значения принимают q_value на недоступных действиях
+        q_value_1 = self.q_function_1.forward(states).detach()
```
Comment (krutovsky-danya) [inline algorithms/soft_actor_critic.py:263]: Нормально ли то, что значения Q функций вычисляется после их градиентного шага?

File: algorithms/soft_actor_critic.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        q_value_2 = self.q_function_2.forward(states).detach()
+
+        min_q_value = torch.min(q_value_1, q_value_2)
+
+        action_probs = self.policy_function.forward_guided(states, guides)
+        action_log_probs = get_nan_safe_log(action_probs, self.small_const)
+        entropy = self.alpha.detach() * action_log_probs
```
Comment (krutovsky-danya) [inline algorithms/soft_actor_critic.py]: Почему бы не вынести это в отдельную функцию?

File: algorithms/soft_actor_critic.py [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+        loss_policy = torch.mean(torch.sum(action_probs * (entropy - min_q_value), dim=1))
```
Comment (krutovsky-danya) [inline algorithms/soft_actor_critic.py]: Кажется все это было уже вычислено в методе выше (только с другим знаком), разве нет?

File: algorithms/soft_actor_critic.py [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+                                        next_states, next_guides)
+        loss_alpha = self._alpha_gradient_step(states, guides)
+
+        return (*losses, loss_alpha)
+
+    def _alpha_gradient_step(self, states, guides):
+        action_probs = self.policy_function.forward_guided(states, guides).detach()
+        action_log_probs = get_nan_safe_log(action_probs, self.small_const)
+
+        loss_alpha = torch.sum(action_probs *
```
Comment (krutovsky-danya) [inline algorithms/soft_actor_critic.py]: Кажется это совпадает полностью с policy_loss, на самом деле даже код особо не нужно менять, просто шагнуть еще по alpha тоже

**Review by krutovsky-danya — CHANGES_REQUESTED** [review]
Очень хорошо, что PR сконцентрирован только на алгоритме!

File: algorithms/soft_actor_critic.py (hunk 1 above)
Comment (Cattharine) [reply algorithms/soft_actor_critic.py:37]: Не думаю. Для меня forward и predict различаются тем, что в forward я могу передать много состояний, а в predict - только одно. Это различие не предполагает отсутствие градиента в predict

File: algorithms/soft_actor_critic.py (hunk 2 above)
Comment (Cattharine) [reply algorithms/soft_actor_critic.py:85]: Тогда уж masked_fill_ (inplace версию masked_fill), так как out-of-place операции копируют тензор, из-за чего тратят больше памяти. Но по сути там уже и так использован masked_fill_, просто запись другая

File: algorithms/soft_actor_critic.py (hunk 3 above)
Comment (Cattharine) [reply algorithms/soft_actor_critic.py]: Увы, да. Судя по всему, np.choice неявно внутри приводит тип к float64, а masked_probs хранит элементы типа float32 (по крайней мере, у меня отрабатывает так), и из-за этого вероятности не суммируются к единице

File: algorithms/soft_actor_critic.py (hunk 6 above)
Comment (Cattharine) [reply algorithms/soft_actor_critic.py]: Будет ли это работать с учетом масок и -torch.inf?

File: algorithms/soft_actor_critic.py (hunk 7 above)
Comment (Cattharine) [reply algorithms/soft_actor_critic.py:263]: Думаю, это нормально. К тому же, так же сделано и в реализации, ссылку на репозиторий с которой я оставила в issue для sac

File: algorithms/soft_actor_critic.py (hunk 9 above)
Comment (Cattharine) [reply algorithms/soft_actor_critic.py]: Нет. В loss для q функций это считается для следующих состояний, в loss для policy это считается для текущих состояний

File: algorithms/soft_actor_critic.py (hunk 10 above)
Comment (Cattharine) [reply algorithms/soft_actor_critic.py]: К сожалению, они не совпадают. В policy_loss ещё учитывается min_q_value, а в alpha_loss учитывается entropy_target

File: algorithms/soft_actor_critic.py (hunk 6 above)
Comment (krutovsky-danya) [reply algorithms/soft_actor_critic.py]: Да, это точно работает с -torch.inf

File: algorithms/soft_actor_critic.py (hunk 7 above)
Comment (krutovsky-danya) [reply algorithms/soft_actor_critic.py:263]: Ух, что-то мне от этого странно, но давай пока так попробуем

Comment (krutovsky-danya) [reply algorithms/soft_actor_critic.py:263]: Кажется там в репозитории активные дискуссии. Давай я заведу issue на эту тему у нас, чтобы поисследовать. Если станет лучше и стабильнее в том репозитории расскажем)

**Review by krutovsky-danya — APPROVED** [review]

File: algorithms/soft_actor_critic.py [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+
+    def _get_probs_and_log_probs(self, states, guides):
+        probs = self.policy_function.forward_guided(states, guides)
+        log_probs = self.policy_function.get_normalized_guided_logits(states, guides)
+        # since probs[~guides] values are supposed to be equal to 0.0
+        # as long as we multiply probs and log_probs
+        # we do not care what is the value of log_probs[~guides]
+        # and also encountering -torch.inf during backward gives us nan
+        # so we'll set log_probs[~guides] to 0.0
+        log_probs = log_probs.masked_fill(~guides, 0.0)
```
Comment (krutovsky-danya) [inline algorithms/soft_actor_critic.py:225]: Тут же не ноль, а -inf. Потому что логарифм нуля - это -inf
